Add comments to explain how to properly use proxy

pulkitsharma07 opened the pull request.

pulkitsharma07 commented:
Please go through the added comments.

pulkitsharma07 commented on BrowserStack TCP Proxy Demo/ConnectionManager/PGConnectionManager.m:
```diff
@@ -1260,11 +1258,11 @@ -(void) addPeer:(int) peerID
 	[socket setDomainID:domainID];
 	
-	[socket setServerIP:@"real.partygaming.com.e7new.com"];
+	[socket setServerIP:@"www.example.org"];
```
Change serverIP and port here.

pulkitsharma07 commented on BrowserStack TCP Proxy Demo/ConnectionManager/PGSocket.m:
```diff
@@ -451,29 +452,40 @@ - (void) dealloc{
+    
+    // inputStream and outputStream variables should be used
+    // no need to create separate socket.
+    CFStreamCreatePairWithSocketToCFHost(NULL, host, 8118, (CFReadStreamRef *)&mInputStream, (CFWriteStreamRef*)&mOutputStream);
```
`host` should be proxyHost
and `8118` should be proxyPort.

pulkitsharma07 commented on BrowserStack TCP Proxy Demo/ViewController.m:
```diff
@@ -91,10 +91,11 @@ - (void)buttonTapped {
             // UPDATE THIS to POKER URL
-            NSString *host = @"real.partygaming.com.e7new.com";
-            UInt32 port = 2147;
+            NSString *host = @"www.example.org";
```
Change to correct values.

Niveditha-ivy commented:
Hi @pulkitsharma07 ,
I don’t see any delegates firing except NSStreamEventOpenCompleted even after applying the above suggested changes. Can you please let me know if you were able to see NSStreamEventHasSpaceAvailable as well?

pulkitsharma07 commented:
Are you able to see handshake logs ? `sendHandShakeMessageToServer`

@Niveditha-ivy  can you push changes to a branch on github, I can just confirm the changes.

Niveditha-ivy commented:
@pulkitsharma07..I am actually unable to commit into recommendations branch.

Niveditha-ivy commented:
@pulkitsharma07  - We don't have write access to push to pulkitsharma07:recommendations.
Please provide access so that will commit my changes.

Niveditha-ivy commented:
Hi @pulkitsharma07 , I am unable to commit the changes due to access issues. Please find the changes in the below PR.

[Changing default url to Poker url] - (https://github.com/Niveditha-ivy/BrowserStackiOS/pull/1/commits/2b56c0a8e876fe22b794287a7c7a49ca433b0b5f)

[Replacing default serverIPs with Poker Domains] - (https://github.com/Niveditha-ivy/BrowserStackiOS/pull/2)

When I verified with the above changes, I was able to see only the NSStreamEventOpenCompleted delegate (Need to fire NSStreamEventHasSpaceAvailable as well).
Let me know if there is any deviation in the above PRs, from the one you have suggested.

pulkitsharma07 commented:
I have invited you as a collaborator, you should be able to push now.

Niveditha-ivy commented:
Thanks for the access. I have pushed my changes please verify.

pulkitsharma07 commented:
I cannot see your commits. Have you pushed to `recommendations` branch ? 
https://github.com/pulkitsharma07/BrowserStackiOS/tree/recommendations

Niveditha-ivy commented:
@pulkitsharma07 , There were some sort if confusion in the Branches here. I guess you should be able to see it now.

commit Hash - 0b925f901d887b6c09bc05b2c97be1cda5215411

pulkitsharma07 commented:
The changes look okay to me. Can you share the log output from xcode ?

Niveditha-ivy commented:
I have attached the log file below, which was when I uploaded the project in the BrowserStack with VPN. 
We cannot execute the project in Xcode due to enabling/finding Proxy issues. 

Note: Please search with the prefix 'BrowserStackLog' in the log file for the project-specific logs. 

[Final_BrowserStackLog_File.log](https://github.com/Entain-Operations-Limited/BrowserStackiOS/files/14430222/Final_BrowserStackLog_File.log)

pulkitsharma07 commented on BrowserStack TCP Proxy Demo/ViewController.m:
```diff
@@ -12,6 +12,104 @@ @interface ViewController ()
+                NSOutputStream *outputStream = (NSOutputStream *)stream;
+                
+                // Prepare the HTTP request
+                NSString *requestString = @"GET http://www.example.org/ HTTP/1.1\r\nHost: www.example.org\r\n\r\n";
```
Sending HTTP request to www.example.org over new socket (via proxy)

pulkitsharma07 commented on BrowserStack TCP Proxy Demo/ViewController.m:
```diff
@@ -283,5 +385,33 @@ -(void)createPokerServerSocket:(NSString *)proxyHost proxyPort:(UInt32)proxyPort
+    
+    NSInputStream *inputStream;
+    NSOutputStream *outputStream;
+    CFStreamCreatePairWithSocketToCFHost(NULL,CFHostCreateWithName(NULL, (CFStringRef)proxyHost), proxyPort, &readStream, &writeStream);
```
this is similar to what poker team is using inside `PGSocket.m`

pulkitsharma07 commented:
Hi @Niveditha-ivy, please check the code I have added. Commit: https://github.com/Entain-Operations-Limited/BrowserStackiOS/pull/1/commits/29b0d4bfd37c83e0767dfd24c823790d2c88f11d

Now, I have skipped Poker code entirely, and tested that all the correct delegate events are firing for request to `www.example.org`

```
2024-03-05 12:43:31.594386+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream event received: NSStreamEventOpenCompleted for stream: InputStream
2024-03-05 12:43:31.594503+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream opened
2024-03-05 12:43:31.594577+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream event received: NSStreamEventOpenCompleted for stream: OuputStream
2024-03-05 12:43:31.594662+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream opened
2024-03-05 12:43:31.594730+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream event received: NSStreamEventHasSpaceAvailable for stream: OuputStream
2024-03-05 12:43:31.594795+0530 BrowserStack TCP Proxy Demo[15390:11409563] Sending HTTP request over stream
2024-03-05 12:43:31.594884+0530 BrowserStack TCP Proxy Demo[15390:11409563] HTTP request sent
2024-03-05 12:43:35.111954+0530 BrowserStack TCP Proxy Demo[15390:11409563] Stream event received: NSStreamEventHasBytesAvailable for stream: InputStream
2024-03-05 12:43:35.112297+0530 BrowserStack TCP Proxy Demo[15390:11409563] Reading HTTP response over stream
2024-03-05 12:43:35.112540+0530 BrowserStack TCP Proxy Demo[15390:11409563] Response received: HTTP/1.1 200 OK
2024-03-05 12:43:35.112540+0530 BrowserStack TCP Proxy Demo[15390:11409563] Response received: HTTP/1.1 200 OK
Accept-Ranges: bytes
Age: 355547
Cache-Control: max-age=604800
Content-Type: text/html; charset=UTF-8
Date: Tue, 05 Mar 2024 07:13:32 GMT
Etag: "3147526947"
Expires: Tue, 12 Mar 2024 07:13:32 GMT
Last-Modified: Thu, 17 Oct 2019 07:18:26 GMT
Server: ECS (dce/26A0)
Vary: Accept-Encoding
X-Cache: HIT
Content-Length: 1256
Proxy-Connection: keep-alive

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-ra
```


Please refer to the method `sendRequestViaProxy` I have added in `ViewController.m`. It is using `CFStreamCreatePairWithSocketToCFHost`, which is also used inside `PGSocket.m` code.

All the correct delegate events are firing. I am able to write HTTP request and also read HTTP response.

Niveditha-ivy commented:
Hi @pulkitsharma07, After the above changes I don’t see any streams connected to Poker server IP.
Here are the consolidated steps we found from the new changes.

1. Open streams with proxyHost & proxyPort
2. Write httpRequest with [www.example.org:80](http://www.example.org/) String through the opened streams
3. If response is OK, -> creating a new socket and streams with proxyHost and proxyPort

Let us know how can we send HandShake request through the streams which are connected to Poker server. Please let me know incase if I miss anything here.

pulkitsharma07 commented:
Yes it is correct, it will not connect to Poker since I have commented that part ([here](https://github.com/Entain-Operations-Limited/BrowserStackiOS/commit/29b0d4bfd37c83e0767dfd24c823790d2c88f11d#diff-5587f3e311b248a3f91a967722e85bba11209633c712853c6dc5599c2f1a33aeR220))

* We connect to proxyHost and proxyPort first (Step 1)
* Then we send `CONNECT www.example.org:80` (Step 2)
  After this the proxy is configured to forward traffic to www.example.org port `80`

* If response is OK, we create new socket to proxyHost:proxyPort and setDelegate on corresponding streams to listen for events.
  * on these streams we can send (outputStream) and receive data (inputStream).
  * in the last commit I have given an example of that.
  * https://github.com/Entain-Operations-Limited/BrowserStackiOS/commit/29b0d4bfd37c83e0767dfd24c823790d2c88f11d#diff-5587f3e311b248a3f91a967722e85bba11209633c712853c6dc5599c2f1a33aeR64 here I am sending a `GET` request to example.org via the outputStream when `NSStreamEventHasSpaceAvailable`. 
  * Similarly when `NSStreamEventHasBytesAvailable` is fired (on inputStream), this means there is response from server, which I am reading here https://github.com/Entain-Operations-Limited/BrowserStackiOS/commit/29b0d4bfd37c83e0767dfd24c823790d2c88f11d#diff-5587f3e311b248a3f91a967722e85bba11209633c712853c6dc5599c2f1a33aeR87


Similar process should be followed for poker use case.

Niveditha-ivy commented:
1. So from the above steps, instead of "www.example.org:80" domain should we configure Poker domain?
2. And once when we receive '200 OK' response, should we create new proxy socket and then after the delegates are fired(NSStreamEventHasSpaceAvailable) should we try sending HandShake to PokerServer?(referring 'sending a GET request to example.org via the outputStream' from the above)

Is this assumption correct @pulkitsharma07?

pulkitsharma07 commented:
yes that's the correct flow.

Using HTTPS might complicate the flow here, so if possible just use plain HTTP request.

pulkitsharma07 commented:
What you mentioned was already happening, in earlier commit. If you check logs for your app [here](https://github.com/Entain-Operations-Limited/BrowserStackiOS/pull/1#issuecomment-1968397797) you can see handshake was being sent to poker. After that it is failing somewhere.

Niveditha-ivy commented:
Hi @pulkitsharma07, 
Hope you are doing well! 
From the current code that we have, our request(HandShake request to poker server) was reaching to server but failing with 'ssl handshake failure'.

There are 2 kinds of ssl verifications here. 
1. While sending the request - Server verifies for SSL/TLS for the request that we send, against the server's configured SSL
2. SSL Validation - Client validates the SSL after receiving response from the server.

As part of our requirement to make the App run on BrowserStack, we have to disable SSL Validation(Point2) and we did that already. But poker server is expecting the HandShake request from the streams for which SSLSettings are configured(Point1). And when we set those config to the streams, streams are getting closed with an error kCFStreamErrorCodeKey=-9847(errSSLHandshakeFail) at client level itself. 

I wanted to know if any proxy/intermediary devices interfere with the SSL handshake process? by any chance? or why we are not able to get the same delegate calls when we set the kCFStreamPropertySSLSettings to the input/outputStreams(in PGSocket connect()). Can you please provide your insights here?

Note: I have cross verified with Android code setup, where we found that the request is being configured with 'TLSV1' version of ssl(Same as in the Normal App).

Niveditha-ivy commented:
@pulkitsharma07 , can you please help us with the above query?
